### ISSUE TITLE
Bottles Next

### ISSUE THREAD

**Issue by mirkobrombin** [bottlesdevs/Bottles#3602]
Latest update:  https://usebottles.com/posts/2024-12-27-rust-libcosmic-next/ 

---

Due to concerns, the issue is currently locked, internal discussions are happening.

**Comment by olumolu** [comment]
Gtk and qt are one of the best option iced is new but heavily worked on by system76 and ubuntu recently using flutter for there desktop and i find it to be smooth enough...
I really don't like the idea to use electron.

**Comment by arjunindia** [comment]
Have you considered [GPUI](https://www.gpui.rs/)? I think it's an interesting project, especially since [zed editor](https://zed.dev/) is built with it and is actually admired by many. It does have issues - lack of documentation, aliasing, needing to use rust, etc. But definitely give it a try!

**Comment by mirkobrombin** [comment]
I will check it as well, ty

**Comment by mirkobrombin** [comment]
Post updated.

**Comment by bitsmurf** [comment]
As written here in this discussion, I would urge you to reconsider the tech stack choice of rust / C#.
There are better alternatives available to achieve what you want:
https://github.com/orgs/bottlesdevs/discussions/3641

**Comment by EddieRingle** [comment]
It seems as though the [most recent blog post with the Rust/libcosmic decision](https://usebottles.com/posts/2024-12-27-rust-libcosmic-next/) follows after only 5 days of [this blog post](https://usebottles.com/posts/2024-12-22-matter-of-technology/) that covers a lot of options that have been evaluated so far (but doesn't seem to mention Rust/libcosmic whatsoever) and leaves the impression that the decision is still largely up in the air. I'm sure there are offline discussions and such that would help explain this, I just wanted to note how, from my outsider perspective, this would have me concerned that this decision was more fueled by hype and the evangelism you see so often from the Rust sphere.

I see Flutter was mentioned (though I'm curious what the specific concerns around code maintainability with larger projects are), along with a few other reactive UI toolkits. Was any thought given to [Kotlin Multiplatform](https://www.jetbrains.com/kotlin-multiplatform/) and Compose UI? It seems like it would meet the stated needs nicely:
  - They're well-tested, and proven in large applications (on Android and elsewhere),
  - actively supported by both corporate efforts (via Google, JetBrains, and others) and open source contributors,
  - in the event there isn't an existing Compose UI component you might need, you have the extremely mature ecosystem of JVM libraries to fall back on until a Kotlin-based solution is ready (at least with regards to the issue of terminal emulators in Flutter called out in this issue and the blog posts).
  - Perhaps this point is one of the more subjective ones, but if you typically use Go then I believe Kotlin will feel nicer than Rust would.
  - KMP's nature and various backend targets (JVM, Native/LLVM, JS, Wasm) means it can be adopted incrementally throughout your proposed architecture for Next (i.e., client, server, and agent) up to and including (if you choose to) making the entire stack Kotlin-based.
  - KMP with Compose would allow you to hedge your bets without tying yourself to a specific underlying toolkit, and ultimately build an app that is as close to "native" (as in native to the given platform/DE) as possible.
    - To expand a bit on this, by building directly against something like libcosmic (or GTK or Qt, for that matter), you're locking yourself into the idiosyncrasies those toolkits bring with them. Linux desktop environments are very diverse, often requiring compatibility shims to maintain consistency (e.g., Qt apps on GNOME, or GTK apps on KDE/Plasma), adding yet another player with libcosmic just adds to that support matrix.
    - Compose-based applications, meanwhile, can be built in a way such that it's possible to use the platform's native toolkit directly without much in the way of duplicated code or other effort. Today, for example, Compose UI on JVM and Wasm would draw to a Skia canvas, but you might also have a target that uses Compose HTML to render out to the DOM. And, while I'm not aware of projects that specifically do this yet as I've not really looked, it would also be possible to emit a tree of GTK widgets, for instance. Another example of this capability is [Mosaic](https://github.com/jakewharton/mosaic), which allows you to build terminal UIs on top of the Compose runtime.

**Comment by bitsmurf** [comment]
@EddieRingle Thank you for sharing a similar concern I have with this choice.
It does indeed seems rushed and driven by hype, rather than going for a tried, tested and mature UI toolkits like GTK / QT.

**Comment by olumolu** [comment]
> @EddieRingle Thank you for sharing a similar concern I have with this choice. It does indeed seems rushed and driven by hype, rather than going for a tried, tested and mature UI toolkits like GTK / QT in a more developer and contribution friendly language.
> 
> From my 23 years of experience in the field of software engineering for multiple platforms, these kind of rushed choices ends up tanking an otherwise great project.

Iced does not offer any accessibility settngs where as gtk4-rs offer.

**Comment by mmstick** [comment]
@olumolu A11y is already implemented in libcosmic, which will continue to improve as COSMIC gets nearer to its 2025 release. A11y settings will be configurable via supported Wayland protocols and cosmic-config. Regarding iced specifically, there is a coordinated effort that will get this merged into iced before the 1.0.0 iced release.

**Comment by TheEvilSkeleton** [comment]
I don't agree with the decision to move to something else, because, as I argued in https://github.com/bottlesdevs/Bottles/issues/1877#issuecomment-1200385617 over two years ago, we should be fixing existing problems, not doing yet another thing. If "we" keep experimenting with different technologies with absurdly high expectations, we'll end up disappointed and/or keep switching technologies; essentially what's happening right now.

We _still_ haven't thought about how Bottles projects/subprojects will split, who's keeping usebottles.com and the `com.usebottles.bottles` app ID on Flathub, etc. What we need is something viable and well coordinated long-term before moving to something else.

Even if we _do_ switch toolkits/frameworks, I'd at least prefer [egui](https://www.egui.rs/), Qt, or even Electron, because they have proper accessibility support. In the case of egui, the maintainer seems to be much more active and responsive with accessibility. That said, sticking with what we currently have is the better approach, especially when it isn't unmaintainable and irreparable, unlike Xorg.

> @olumolu A11y is already implemented in libcosmic, which will continue to improve as COSMIC gets nearer to its 2025 release. A11y settings will be configurable via supported Wayland protocols, desktop integrations, and cosmic-config. Regarding iced specifically, there is a coordinated effort that will get this merged into iced before the 1.0.0 iced release. Though the Orca support under Wayland currently requires patches to Orca that are only in the Pop!_OS 24.04 Alpha at the moment.

"A11y is already implemented" doesn't have much value to me, because implementing accessibility is _much_ easier than maintaining it. It needs to be properly nurtured and maintained (financially, diversely, etc.) because it is a malleable topic that is actively being researched. Accessibility is still largely nonexistent in libcosmic, so what @olumolu's point is fair and irrefutable: there is no accessibility; it doesn't matter if it's being worked on or not.

**Comment by acerspyro** [comment]
> @olumolu A11y is already implemented in libcosmic, which will continue to improve as COSMIC gets nearer to its 2025 release. A11y settings will be configurable via supported Wayland protocols, desktop integrations, and cosmic-config. Regarding iced specifically, there is a coordinated effort that will get this merged into iced before the 1.0.0 iced release. Though the Orca support under Wayland currently requires patches to Orca that are only in the Pop!_OS 24.04 Alpha at the moment.

That's not a good reason to throw the baby out with the bathwater.

I strongly believe this is wasted development time that could be spent on more important items, like actually implementing A11Y in the existing version of Bottles, which would likely be a quicker affair than switching to libcosmic.

This also wastes development time that was already spent on building the GTK4 UI.

It also does not fit well with the direction of Vanilla OS, which comes with Bottles pre-installed and uses GNOME with Libadwaita applications.

I agree with @TheEvilSkeleton and @bitsmurf here, this smells like a "I'm bored, let's hop on the next hypetrain" decision rather than a technically sound one.

Besides, even if its A11Y support gets completed in 2025 as promised, libcosmic is not mature software and may introduce a ton of unforeseen issues for a project as large as this one.

**Comment by jamiethecat** [comment]
I'd genuinely like you to consider Flutter specifically here if you want to expand beyond Linux.

For me it's mostly a battle between native options, since there's definitely enough of those for a project like Bottles (web tech would still be better than libcosmic for various reasons though). It gives you:

- A strong official backing
- Plenty of official and third party packages
- A decent language (it's not perfect, but a lot more readable than Rust)
- More contributor potential since it's an established project

**Comment by acerspyro** [comment]
> I'd genuinely like you to consider Flutter specifically here if you want to expand beyond Linux.
> 
> For me it's mostly a battle between native options, since there's definitely enough of those for a project like Bottles (web tech would still be better than libcosmic for various reasons though). It gives you:
> 
>     * A strong official backing
> 
>     * Plenty of official and third party packages
> 
>     * A decent language (it's not perfect, but a lot more readable than Rust)
> 
>     * More contributor potential since it's an established project

It would likely require a complete re-write of Bottles to Dart, unless you want to use a 9 month old Python shim project that hasn't seen a 1.0 release yet: https://flet.dev/

**Comment by acerspyro** [comment]
> > It would likely require a complete re-write of Bottles to Dart
> 
> Oh absolutely, it would be a huge effort, but I see this as the best direction if Bottles wants to expand beyond Linux. If it stays Linux-only software, the current stack is fine as is ^^

The current stack is macOS-compatible, Libadwaita runs on macOS. See Zrythm for an example.

**Comment by jamiethecat** [comment]
> The current stack is macOS-compatible, Libadwaita runs on macOS. See Zrythm for an example.

I have experience with GTK and Libadwaita on macOS, and saying it "runs" is correct, it just doesn't run well and gets broken here and there.

**Comment by acerspyro** [comment]
> > The current stack is macOS-compatible, Libadwaita runs on macOS. See Zrythm for an example.
> 
> I have experience with GTK and Libadwaita on macOS, and saying it "runs" is correct, it just doesn't run well and gets broken here and there.

What do you think would be the time cost of fixing macOS-specific GTK4 issues here and there versus throwing away the whole stack and starting over?

**Comment by mirkobrombin** [comment]
It seems like two key points are being completely ignored:  

1. The current codebase does NOT allow us to achieve the new Next experience, which is the core feature of the entire Next project. This would mean rewriting the entire frontend and part of the backend. Since I’m not happy with the current structure or the stack, this step is necessary.  

2. The current codebase still uses subprocesses to handle everything. Every single operation is a massive and complex command generated through equally complex classes, starting with WineCommand, which is a complete mess to untangle. Next aims to control operations from inside the prefix, using native Windows interactions and accessing them through simple, clean, and clear gRPC instructions (or similar, since we’re still finalizing the protocol).  

I understand the concerns about accessibility, and I respect them, but these were issues even with Electron, Tauri, and Wails.  

Regarding Flutter, I already linked an issue that explains one of its many package-related problems. I’ll also add my fear of seeing Google kill yet another project. In some way, I feel more confident and reassured by System76 than by Google nowadays.  

From a programming language perspective, learning Dart isn’t as advantageous for me or others, while Rust is becoming more and more widely used and probably has one of the largest communities today.  

Another point that is being completely ignored is that Next is multi-client. What matters is the server (backend), not the client.  

There will be a GTK client, like the current one, implementing the Next mode and using the new backend. There will also be a libcosmic client, because right now it’s a bet I honestly believe in.  

Neither team, whether working on Next or the current version, is forced to work on the libcosmic version. The GUI clients will be the last to be developed, since the very first client will be CLI. Once all the core components like WineBridge, Server, and CLI are finished, the teams can decide how to move forward, either collaborating on the GTK version or splitting up to work on both.  

I also read a lot of frustration in the previous comments, almost as if the Next project is just about rewriting everything without a reason. I invite everyone to read the blog posts, especially the first one.  

About the tech stack and the many options discussed, I’ve spent my time, unpaid, evaluating ALL the possibilities, documenting everything as much as possible, and explaining the direction, including this latest change.  

I’ve listened to the community, those against web-based solutions, against Go, against GTK, against this and that, and I chose to set aside my own interest in GTK and libadwaita, leaving their management to the rest of the team. I personally decided to bet on something new, recognizing the effort System76 is putting into it, which is honestly one of the few Linux and Open Source companies I trust today.

**Comment by jamiethecat** [comment]
> What do you think would be the time cost of fixing macOS-specific GTK4 issues here and there versus throwing away the whole stack and starting over?

You'd have to constantly keep up with GObject and practically co-maintain it which is a lot if work. Since the entire structure of Bottles is being reworked, I'd say Flutter would actually *save* time as I see the future being more stable (there is real focus on cross-platform compatibility).

**Comment by TheEvilSkeleton** [comment]
> The current codebase does NOT allow us to achieve the new Next experience, which is the core feature of the entire Next project. This would mean rewriting the entire frontend and part of the backend. Since I’m not happy with the current structure or the stack, this step is necessary.

Exactly, this is doable because we're rewriting parts of the backend. We could throw away the frontend and start again there, but the backend would be here to stay.

> I understand the concerns about accessibility, and I respect them, but these were issues even with Electron, Tauri, and Wails.

That's why I didn't mention Tauri and Wails. And no, unlike Iced/libcosmic, there is accessibility support in Electron.

> Another point that is being completely ignored is that Next is multi-client. What matters is the server (backend), not the client.

I mean, that's not my point. We're rewriting and rethinking an entire project of 15 thousand lines of code. That's gonna take way more than a year to develop, and even more to polish it. Furthermore, we still haven't thought about how the projects will be split and who's keeping the current app ID (I call dibs).

> I’ve listened to the community, those against web-based solutions, against Go, against GTK, against this and that, and I chose to set aside my own interest in GTK and libadwaita, leaving their management to the rest of the team. I personally decided to bet on something new, recognizing the effort System76 is putting into it, which is honestly one of the few Linux and Open Source companies I trust today.

I suppose that's fair, I don't want to stop you from enjoying maintaining Bottles again. The main issue is that the entire decision was rushed, because we haven't thought about what we currently have. I'm still of the opinion to rewrite parts of the backend and adapt the frontend to it, rather than rewriting everything from scratch.

**Comment by mirkobrombin** [comment]
> Exactly, this is doable because we're rewriting parts of the backend. We could throw away the frontend and start again there, but the backend would be here to stay.

No, because the whole backend is based on subprocess and commands interpolation. Remove that approach then remove the whole backend. What stay is basically some utils.

> ..there is accessibility support in Electron.

Sure but you still have to support and implement that in the toolkit you are using in the web app, the fact the tech supports accessibility, does not mean the software gets it for free. Especially for the web.

> I mean, that's not my point. We're rewriting and rethinking an entire project of 15 thousand lines of code. That's gonna take way more than a year to develop, and even more to polish it. Furthermore, we still haven't thought about how the projects will be split and who's keeping the current app ID (I call dibs).

This is why I am creating a whole new team, to not steal time from the main one. Before ever thinking about the app ID, the Next team should put everything on paper.

> I suppose that's fair, I don't want to stop you from enjoying maintaining Bottles again. The main issue is that the entire decision was rushed, because we haven't thought about what we currently have. I'm still of the opinion to rewrite parts of the backend and adapt the frontend to it, rather than rewriting everything from scratch.

Legacy and Next coexist, for a very long time. I don't know why this doesn't seem clear. We can continue to improve it, rewrite parts... Next is a project in itself. We'll consider giving it a whole new name but it would be stupid to throw away all the work I've done since 2017 on social media and gaining popularity.

**Comment by jamiethecat** [comment]
> Regarding Flutter, I already linked an issue that explains one of its many package-related problems.

I think that's unfair. Do you want me to go across COSMIC apps to show you all the libcosmic issues? Because I certainly can, to name a few:

- doesn't follow the window controls
- doesn't follow system style preference
- adaptivity is broken
- lack of animations

And that's only on Linux!

> learning Dart isn’t as advantageous for me or others

It will take you approximately 10 minutes to learn it if you already have programming experience.

I'm sorry to say but, this feels like a rushed decision that was once again not talked over with other contributors (I can't know this, of course) :(

**Comment by bitsmurf** [comment]
> while Rust is becoming more and more widely used and probably has one of the largest communities today
Except rust is poorly designed and has a steep learning curve - not really inviting for contributions.
And to be frank, it isn't even suited for this kind of project either.
The majority of rust is driven by hype, rather than practicality.

This conversation is important to have, and also why I brought up Lazarus / Free Pascal as a viable contender to drive the bottles-next project forward.
You get to develop and design beautiful GUIs while being able to target Cocoa on macOS, GTK or QT on Linux,, and the Win32 API under Wine for the wine bridge.

Use the right tool for the job.

**Comment by TheEvilSkeleton** [comment]
> Sure but you still have to support and implement that in the toolkit you are using in the web app, the fact the tech supports accessibility, does not mean the software gets it for free. Especially for the web.

No, semantic HTML is a thing, so is the accessibility tree. But okay, since Electron doesn't come with accessibility for free, in that case, we can ignore Electron, and we'll still have all the other options available.

> Legacy and Next will coexist, for a very long time. I don't know why this doesn't seem clear. We can continue to improve it, rewrite parts... Next is a project in itself. We can consider giving it a new name but it would be stupid to throw away all the work we've done since 2017 on social media to give popularity to the project.

It's clear, but I just don't agree. I'm also not comfortable using "we" in these articles, because I didn't agree with these decisions :/

**Comment by TheEvilSkeleton** [comment]
Please try to understand that tons of decisions we took as a team in the past, all the discussions and controversies that happened in this project, put me in a position where I'm considered to be "the person who works on Bottles". Every decision "we" take, I'm involved in, whether I like it or not.

**Comment by bitsmurf** [comment]
@TheEvilSkeleton I can only speak for myself, but I want to say thank you for everything you've done so far - and continues to do for Bottles. It haven't gone unnoticed 🙂

**Comment by mirkobrombin** [comment]
This conversation is going nowhere, the only thing that comes across (unfairly) is that I rush decisions making me look bad. I've proposed again an internal discussion, hoping to receive feedback from the team this time, we'll reopen the issue when it's no longer necessary for me to appear as the bad guy.

**Comment by mirkobrombin** [comment]
I was attempting to clean up the discussion by simply hiding off topic comments and outdated, since internal discussions are happening, but I prefer to close the discussion and make a new article when that discussion is ended.
